packages/block-editor/src/components/provider/index.js

```diff
@@ -120,6 +126,12 @@ class BlockEditorProvider extends Component {
 				blocks = newBlocks;
 				isPersistent = newIsPersistent;
 
+				// Selection must be updated first, so it is recorded in history when the content change happens.
+				onChangeSelection( [
```

Comment:
I think we should also do the same for `onChange` and `onInput`, but this callback should be optional.

It should also be documented in the component's README.md.

https://github.com/WordPress/gutenberg/blob/65444a5ca1b40a49760c4a26203be1dac99f5a70/packages/block-editor/src/components/provider/README.md